### ISSUE TITLE
Make subprocess timeouts configurable per project and per tool, and report them on timeout

### ISSUE THREAD

**Issue by MarcusJellinghaus** [MarcusJellinghaus/mcp-code-checker#219]
## Problem

Every checker subprocess runs under a timeout nobody chose. `execute_command` defaults to `timeout_seconds: int = 120` (`CommandOptions` in mcp-coder-utils). **Nine programs** run at 120s: mypy and pylint pass a hardcoded `timeout_seconds=120`, and seven — ruff, bandit, vulture, tach, lint-imports, black, isort — inherit the default silently, never passing the argument at all. Only pytest differs, at 300s. None of it is reachable from configuration.

On a large project this makes `run_mypy_check` unusable. A cold-cache strict run over `mcp-coder` (729 files) exceeded 120s at every scope tried — default targets, `["src"]`, a single leaf package, and `follow_imports="silent"` — and returned only:

```
Mypy execution failed: Process timed out after 120 seconds
```

Narrowing `target_directories` does not help, because mypy still type-checks everything the targeted modules import.

| Checker | Timeout | Configurable? | Location |
|---|---|---|---|
| rope refactoring | 120s | Yes — `--refactoring-timeout` | `main.py:97` → `server.py` → `RefactoringTools` |
| mypy | 120s | No — hardcoded literal | `code_checker_mypy/runners.py:140` |
| pylint | 120s | No — hardcoded literal | `code_checker_pylint/runners.py:100` |
| pytest | 300s | Partly — a defaulted parameter, not exposed by the MCP tool | `code_checker_pytest/runners.py:91` |
| ruff, bandit, vulture, tach, lint-imports, black, isort | 120s | No — inherited silently from `execute_command` | never passed at all |

A second defect compounds this: no checker reports a timeout correctly. See the decision below.

### Two settings, one shared and one per tool

`check-timeout` applies to every tool; a per-tool value overrides it.

Resolution chain: **tool argument → pyproject per-tool → pyproject shared → `--check-timeout` → built-in**

### `check-timeout` covers all ten MCP tools, pytest included

Previously scoped to nine. A shared setting that skips the slowest tool is not shared: `check-timeout = 60` would leave pytest at 300s, and `check-timeout = 600` would raise everything except the tool that most needs it. pytest's built-in default stays 300s, so the two only differ when nothing is configured.

### Per-tool keys are named after programs, not MCP tools

The ten MCP tools and the ten programs they run are not 1:1:

- `run_format_code` is one tool running two programs (isort, then black)
- ruff is one program behind two tools, and `run_ruff_fix` spends its timeout twice (a pre-check call, then the apply call)

So: `black-timeout` and `isort-timeout` separately, one shared `ruff-timeout`. A key bounds **one run of one program**, not one tool call, so a single tool call can spend more than one budget — the Documentation section lists those cases.

The alternatives — a `format-timeout` split across two programs, or halving a configured value — make the number in the config file not the number in effect.

### Every checker must report its timeout

Folded into this issue rather than split, because each fix is a branch reordering in a function this issue already opens.

On timeout `execute_command` sets `timed_out=True` **and** `execution_error="Process timed out after N seconds"`, with empty stdout/stderr. Every checker mishandles one or the other:

- **Five shadow their own `timed_out` branch** by testing `execution_error` first, making it unreachable dead code — mypy (`:150` before `:156-161`), pylint (`:104` before `:124-130`), bandit (`:65` before `:73-76`), ruff (all three call sites: `:73`, `:121`, `:149`), pytest (`:225` raises `RuntimeError` before the `TimeoutError` at `:231-242`). These report the generic `execution_error` text, which is at least true.
- **Five ignore both fields** and report something false:
  - `code_checker_tach/runners.py` returns `"tach check passed (no output)."` — **a killed check reports green**
  - `code_checker_vulture/runners.py` returns `"vulture produced no output."` — reads as "no dead code found", **also a false pass**
  - `code_checker_lint_imports/runners.py` returns `"ERROR: lint-imports output could not be parsed"` — wrong cause
  - `formatter/black_runner.py` and `isort_runner.py` return `success=False` with empty output — no reason given

The observed `Mypy execution failed: Process timed out after 120 seconds` comes from the generic path (`code_checker_mypy/reporting.py:116` prefixes `result.error`), which is proof the hand-written message at mypy `:160` never fires. So the fix is uniform: check `timed_out` before `execution_error`, and report the limit actually used.

Raising the limit does not fix this, and making timeouts configurable means more configurations reach it.

### A malformed `pyproject.toml` raises

`utils/project_config.py` holds two conflicting idioms: `get_target_directories` raises `ValueError(f"Invalid pyproject.toml: {exc}")` (`:54`), while `check_line_length_conflicts` swallows `TOMLDecodeError` and returns `[]` (`:157-160`). Timeout resolution follows `get_target_directories` and raises — a syntactically broken `pyproject.toml` fails every tool call, and failing loudly is the only honest answer.

Consequence: combined with per-call resolution, this newly makes `run_tach_check` and `run_lint_imports_check` fail on a broken `pyproject.toml`; they read no project config at all today. The swallowing idiom in `check_line_length_conflicts` is a pre-existing inconsistency that this issue does not resolve.

### Two surfaces, asymmetric

`[tool.mcp-tools-py]` in the checked project's `pyproject.toml` carries both shared and per-tool values. The CLI carries **only** `--check-timeout` — one blanket server-level policy.

This removes the ambiguous case where a per-tool CLI flag would fight a shared pyproject value, and keeps the mcp-config follow-up to a single `ParameterDef`.

```toml
[tool.mcp-tools-py]
check-timeout = 300
mypy-timeout = 600
pytest-timeout = 900
```

### Positive integers only

`0` and negatives are rejected with a clear error. `0 = disabled` is decided against: an unbounded subprocess in an MCP server is an unrecoverable hang, since nothing else will reap it and the tool call simply never returns — which is the failure mode this issue exists to remove. Document that a large value approximates "never".

### Per-call `timeout_seconds` on `run_pytest_check` and `run_mypy_check` only

These are the two tools with genuinely variable runtimes, and it gives an agent a way to retry after a timeout instead of being stuck until a human edits config. The remaining tools keep clean signatures — a per-call timeout on ruff is noise in every tool schema the agent loads.

### Built-in defaults unchanged

120s, 300s for pytest. Configuration is opt-in.

*Consequence: the mcp-coder case above does not improve until that project sets the config.*

## Implementation notes

- **Thirteen subprocess invocations, not ten.** One per program, plus two extra ruff calls (`code_checker_ruff/runners.py:68` for `run_ruff_check`, then `:116` pre-check and `:144` apply for `run_ruff_fix`) and one extra pytest call (`code_checker_pytest/runners.py:296`, the retry after the plugin install, passing the same `timeout_seconds`).
- **Two subprocesses are deliberately excluded**: the tool-availability probe (`server.py:222`, 10s, `python -m <tool> --version`) and the pytest-json-report pip install (`code_checker_pytest/runners.py:264`, 60s). Neither runs user code; both stay hardcoded.
- **`RefactoringTools` stays outside `check-timeout`.** It is the only subprocess-running tool group that does not receive the server object — it takes `project_dir` and `timeout` directly — and it already has `--refactoring-timeout`.
- **The hardcoded "120 seconds" messages are dead, not wrong.** mypy `runners.py:160` and pylint `runners.py:128` sit behind the unreachable branches described above, so today they cannot lie. Reordering the branches makes them live, and only then must they report the value actually used.
- **Plumbing is available.** `ToolServer.__init__` passes `self` to `CheckerTools(self)` and `FormatterTools(self)`, so `server.project_dir` reaches every checker and formatter call site; adding `server.check_timeout` follows. Depth is one hop for bandit, ruff, tach, vulture and lint-imports; two for mypy (`get_mypy_prompt` → `run_mypy_check`), pylint (`get_pylint_prompt` → `get_pylint_results`), pytest (`check_code_with_pytest` → `run_tests`) and black/isort (`run_format_code` → `run_black`/`run_isort` via `_STEP_RUNNERS`).
- **The two-keys-one-tool decision becomes concrete at exactly one call site.** `formatter/runner.py:60-63` dispatches both step runners through a single positional signature, `runner(python_executable, target_dirs, project_root, check_only)`, so separating `black-timeout` from `isort-timeout` needs a per-step lookup there.
- **CLI precedent to mirror**: `main.py` argparse (`type=int, default=120`) → `create_server(refactoring_timeout=...)` → `ToolServer.__init__` stores it → passed at tool-group registration.
- **Resolution runs per call**, not at startup, following the existing `resolve_target_directories(str(server.project_dir), ...)` precedent — timeout edits then take effect without a server restart, at the cost of one TOML parse per tool call.
- **Reuse the `tomllib` load from `utils/project_config.py`.** Do **not** copy its fallback-with-warning idiom (`get_target_directories` collecting into `TargetDirs.warnings`) — that warns because a missing section is unexpected, whereas for timeouts a missing section is the normal case, so it would warn on every tool call for every project.
- **Validation is needed on three surfaces**: argparse, the pyproject value, and the per-call `timeout_seconds` argument. An agent passing `timeout_seconds=0` to `run_mypy_check` is the likeliest of the three.
- `CommandOptions.timeout_seconds` is `int` with no `None` branch. This is why positive-only.
- pytest already has the parameter (`runners.py:91`, `:479`) plumbed to the call at `:215`; only the tool surface is missing.
- **The effective limit is `min(server timeout, harness timeout)`.** A calling agent's watchdog can cut a tool call short regardless of this setting. Not solved here; recorded so a large value is not mistaken for a guarantee.
- **Test impact is small.** Most runner tests assert only the command list, so a new keyword argument is invisible to them. Three strict assertions need updating: `tests/test_code_checker/test_runners.py:221-233` (positional `assert_called_once_with` on `run_tests` including `300,  # timeout_seconds`), `tests/test_server_params.py:74-85`, `tests/test_checker_tools.py:478`. `tests/test_refactoring/test_rope_tools.py:432-457` is the pattern for the new `--check-timeout` tests, and `tests/test_project_config.py` is the home for resolution-chain tests.
- **Do not copy the existing `test_timeout` tests.** `make_command_result` (`tests/conftest.py:6-20`) leaves `execution_error=None` by default, so `make_command_result(timed_out=True)` builds a state `execute_command` never produces — which is why the bandit and ruff `test_timeout` tests pass today against the dead branches. Reusing that mock shape for the new tach, vulture and lint-imports tests would give green tests over an unfixed production path. Every new or updated timeout test must set `timed_out=True` **together with** `execution_error`.

## Documentation (required)

- `argparse` help for `--check-timeout`
- README §Optional Parameters — new row, plus **backfill `--refactoring-timeout` and `--vulture-whitelist`**, which have argparse help but appear in no table
- `docs/pyproject-configuration.md` — new `[tool.mcp-tools-py]` section with the full precedence chain. The file is titled "Pylint Configuration via `pyproject.toml`" but already covers more than pylint (it has a "Target directory auto-detection" section at line 77), so it needs retitling first. This would also be the first config section mcp-tools-py itself owns — today it only reads sections owned by other tools.
- The ten per-tool keys, all valid TOML bare keys with no collision between them: `mypy-timeout`, `pylint-timeout`, `pytest-timeout`, `ruff-timeout`, `bandit-timeout`, `vulture-timeout`, `tach-timeout`, `lint-imports-timeout`, `black-timeout`, `isort-timeout`
- The three tool calls that can spend more than one budget: `run_format_code` up to `black-timeout + isort-timeout`; `run_ruff_fix` up to 2× `ruff-timeout`; `run_pytest_check` up to 2× `pytest-timeout` plus the 60s install, when the pytest-json-report plugin is missing and the run is retried
- One sentence on `pytest-timeout` colliding by name with the well-known PyPI plugin. There is no actual TOML clash (that plugin reads `[tool.pytest.ini_options] timeout`), but the similarity is worth calling out.
- That a malformed `pyproject.toml` now fails every tool call, including `run_tach_check` and `run_lint_imports_check`
- Docstrings for the two new tool arguments

## Dependencies / references

- **mcp-config**: `--check-timeout` is inert for normal users until mcp-config adds a matching `ParameterDef` to `MCP_CODE_CHECKER` (`src/mcp_config/servers.py:493`), where this server is registered under the name `mcp-code-checker`. Confirmed: neither `refactoring-timeout` nor `vulture-whitelist` has a `ParameterDef` there, which is why both are unreachable today. The pyproject surface works immediately and has no such dependency.
- Supersedes #73 and #42, both already closed.
- Cross-repo intent from #73: mcp-workspace and mcp-tools-sql inherit the same 120s `execute_command` default and could adopt this pattern.

## Out of scope

mypy cold-cache recovery — see #225.

**Comment by MarcusJellinghaus** [comment]
## Follow-up: sharper root cause — it's the *cold* cache, and the cache is unreachable from inside MCP

Further investigation on the same project refines the problem statement above. The 120s cap is not too low in general — it is only fatal on a cold cache, and the tool provides no way to warm one.

### Measurements

| Run | Time |
|---|---|
| mypy, cold cache, server's flag set | **> 120s** (killed, every attempt) |
| mypy, warm cache, same flag set | **~22s** |
| `run_mypy_check` via MCP, after warming | **completes — "No type errors found"** |

So the tool works fine once the cache exists. Everything in the original report was a cold-cache symptom.

### The trap: the cache is keyed on the option set

My first warming attempt ran mypy with `--strict` and did **not** fix the timeout, which was initially confusing. The reason is that `STRICT_FLAGS` (`code_checker_mypy/runners.py:18-32`) is not equivalent to `--strict`:

```python
STRICT_FLAGS = [
    "--strict",
    "--warn-redundant-casts",     # implied by --strict
    "--warn-unused-ignores",      # implied by --strict
    "--warn-unreachable",         # NOT implied by --strict
    ...
]
```

`--warn-unreachable` is not part of `--strict`. mypy keys its incremental cache on the effective option set, so a cache built with `--strict` is discarded by the server's invocation and it rebuilds from scratch — timing out again. Only after re-warming with the server's **exact** flag list (including `--output json --no-color-output --follow-imports normal` and the full `STRICT_FLAGS`) did `run_mypy_check` start completing.

### Why this is worse than a tunable timeout

Once a project is large enough that a cold run exceeds 120s, `run_mypy_check` becomes **permanently unusable and unrecoverable from inside MCP**:

- every call rebuilds from cold and is killed before it can write a complete cache
- the only escape is running mypy outside the server, which defeats the point in an MCP-only workflow (and may be disallowed by a project's own tooling rules)
- a fresh clone, a CI runner, a cleared `.mypy_cache`, a dependency upgrade, or any change to `STRICT_FLAGS` puts it straight back into that state
- the failure surfaces as `Mypy execution failed: Process timed out after 120 seconds`, which gives no hint that a cache is the issue

### Suggestions, in addition to the configurable timeout

1. **Better diagnostics.** On timeout, say whether the cache directory was populated and mention that a cold run may exceed the limit. The current message sends people looking for the wrong problem.
2. **A cache-priming path.** Either a dedicated tool (`warm_mypy_cache`) or an opt-in longer budget for a first run, so the server can bootstrap itself without an external mypy invocation.
3. **Publish the exact flag set** in the tool docstring, so anyone warming the cache externally can match it. Getting this wrong is silent — you simply keep timing out.
4. **Consider whether `--warn-unreachable` belongs in `STRICT_FLAGS`.** It is deliberately excluded from `--strict` upstream because it produces false positives with `TYPE_CHECKING`, platform checks, and version guards. Not a bug, but worth being a documented choice given it is what breaks naive cache warming.

Point 2 is arguably the real fix: with it, the timeout matters far less, because the expensive run happens once under a budget that suits it.
